<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Visualization of single-cell RNA seq data &mdash; vitessce  documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/stylesheet.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "document", "processHtmlClass": "math|output_area"}}</script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> vitessce
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../widget_examples.html">Widget examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../data_examples.html">Data preparation examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api_config.html">View config API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api_data.html">Data preparation APIs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../screenshots.html">Screenshots</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">vitessce</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Visualization of single-cell RNA seq data</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/vitessce/vitessce-python/blob/master/docs/notebooks/web_app_brain.ipynb" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="Visualization-of-single-cell-RNA-seq-data">
<h1>Visualization of single-cell RNA seq data<a class="headerlink" href="#Visualization-of-single-cell-RNA-seq-data" title="Permalink to this headline"></a></h1>
<section id="1.-Import-dependencies">
<h2>1. Import dependencies<a class="headerlink" href="#1.-Import-dependencies" title="Permalink to this headline"></a></h2>
<p>We need to import the classes and functions that we will be using from the corresponding packages.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">from</span> <span class="nn">os.path</span> <span class="kn">import</span> <span class="n">join</span>
<span class="kn">from</span> <span class="nn">urllib.request</span> <span class="kn">import</span> <span class="n">urlretrieve</span>
<span class="kn">from</span> <span class="nn">anndata</span> <span class="kn">import</span> <span class="n">read_h5ad</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>

<span class="kn">from</span> <span class="nn">vitessce</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">VitessceConfig</span><span class="p">,</span>
    <span class="n">Component</span> <span class="k">as</span> <span class="n">cm</span><span class="p">,</span>
    <span class="n">CoordinationType</span> <span class="k">as</span> <span class="n">ct</span><span class="p">,</span>
    <span class="n">AnnDataWrapper</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="2.-Download-the-data">
<h2>2. Download the data<a class="headerlink" href="#2.-Download-the-data" title="Permalink to this headline"></a></h2>
<p>For this example, we need to download a dataset from the COVID-19 Cell Atlas <a class="reference external" href="https://www.covid19cellatlas.org/index.healthy.html#habib17">https://www.covid19cellatlas.org/index.healthy.html#habib17</a>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="s2">&quot;data&quot;</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">adata_filepath</span> <span class="o">=</span> <span class="n">join</span><span class="p">(</span><span class="s2">&quot;data&quot;</span><span class="p">,</span> <span class="s2">&quot;habib17.processed.h5ad&quot;</span><span class="p">)</span>
<span class="n">urlretrieve</span><span class="p">(</span><span class="s1">&#39;https://covid19.cog.sanger.ac.uk/habib17.processed.h5ad&#39;</span><span class="p">,</span> <span class="n">adata_filepath</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="3.-Load-the-data">
<h2>3. Load the data<a class="headerlink" href="#3.-Load-the-data" title="Permalink to this headline"></a></h2>
<p>Note: this function may print a <code class="docutils literal notranslate"><span class="pre">FutureWarning</span></code></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">read_h5ad</span><span class="p">(</span><span class="n">adata_filepath</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="3.1.-Preprocess-the-Data-For-Visualization">
<h2>3.1. Preprocess the Data For Visualization<a class="headerlink" href="#3.1.-Preprocess-the-Data-For-Visualization" title="Permalink to this headline"></a></h2>
<p>This dataset contains 25,587 genes. In order to visualize it efficiently, we convert it to CSC sparse format so that we can make fast requests for gene data. We also prepare to visualize the top 50 highly variable genes for the heatmap as ranked by dispersion norm, although one may use any boolean array filter for the heatmap.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">top_dispersion</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;dispersions_norm&quot;</span><span class="p">][</span>
    <span class="nb">sorted</span><span class="p">(</span>
        <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;dispersions_norm&quot;</span><span class="p">])),</span>
        <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;dispersions_norm&quot;</span><span class="p">][</span><span class="n">k</span><span class="p">],</span>
    <span class="p">)[</span><span class="o">-</span><span class="mi">51</span><span class="p">:][</span><span class="mi">0</span><span class="p">]</span>
<span class="p">]</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;top_highly_variable&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;dispersions_norm&quot;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">top_dispersion</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="4.-Create-the-Vitessce-widget-configuration">
<h2>4. Create the Vitessce widget configuration<a class="headerlink" href="#4.-Create-the-Vitessce-widget-configuration" title="Permalink to this headline"></a></h2>
<p>Vitessce needs to know which pieces of data we are interested in visualizing, the visualization types we would like to use, and how we want to coordinate (or link) the views.</p>
<section id="4.1.-Instantiate-a-VitessceConfig-object">
<h3>4.1. Instantiate a <code class="docutils literal notranslate"><span class="pre">VitessceConfig</span></code> object<a class="headerlink" href="#4.1.-Instantiate-a-VitessceConfig-object" title="Permalink to this headline"></a></h3>
<p>Use the <code class="docutils literal notranslate"><span class="pre">VitessceConfig(name,</span> <span class="pre">description)</span></code> constructor to create an instance.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">vc</span> <span class="o">=</span> <span class="n">VitessceConfig</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;Habib et al&#39;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s1">&#39;COVID-19 Healthy Donor Brain&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="4.2.-Add-a-dataset-to-the-VitessceConfig-instance">
<h3>4.2. Add a dataset to the <code class="docutils literal notranslate"><span class="pre">VitessceConfig</span></code> instance<a class="headerlink" href="#4.2.-Add-a-dataset-to-the-VitessceConfig-instance" title="Permalink to this headline"></a></h3>
<p>In Vitessce, a dataset is a container for one file per data type. The <code class="docutils literal notranslate"><span class="pre">.add_dataset(name)</span></code> method on the <code class="docutils literal notranslate"><span class="pre">vc</span></code> instance sets up and returns a new dataset instance.</p>
<p>Then, we can call the dataset’s <code class="docutils literal notranslate"><span class="pre">.add_object(wrapper_object)</span></code> method to attach a “data wrapper” instance to our new dataset. For example, the <code class="docutils literal notranslate"><span class="pre">AnnDataWrapper</span></code> class knows how to convert AnnData objects to the corresponding Vitessce data types.</p>
<p>Dataset wrapper classes may require additional parameters to resolve ambiguities. For instance, <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> objects may store multiple clusterings or cell type annotation columns in the <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> dataframe. We can use the parameter <code class="docutils literal notranslate"><span class="pre">cell_set_obs_cols</span></code> to tell Vitessce which columns of the <code class="docutils literal notranslate"><span class="pre">obs</span></code> dataframe correspond to cell sets.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">dataset</span> <span class="o">=</span> <span class="n">vc</span><span class="o">.</span><span class="n">add_dataset</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;Brain&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">add_object</span><span class="p">(</span><span class="n">AnnDataWrapper</span><span class="p">(</span>
        <span class="n">adata</span><span class="p">,</span>
        <span class="n">mappings_obsm</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;X_umap&quot;</span><span class="p">],</span>
        <span class="n">mappings_obsm_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;UMAP&quot;</span><span class="p">],</span>
        <span class="n">cell_set_obs</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CellType&quot;</span><span class="p">],</span>
        <span class="n">cell_set_obs_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Cell Type&quot;</span><span class="p">],</span>
        <span class="n">expression_matrix</span><span class="o">=</span><span class="s2">&quot;X&quot;</span><span class="p">,</span>
        <span class="n">matrix_gene_var_filter</span><span class="o">=</span><span class="s2">&quot;top_highly_variable&quot;</span>
    <span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="4.3.-Add-visualizations-to-the-VitessceConfig-instance">
<h3>4.3. Add visualizations to the <code class="docutils literal notranslate"><span class="pre">VitessceConfig</span></code> instance<a class="headerlink" href="#4.3.-Add-visualizations-to-the-VitessceConfig-instance" title="Permalink to this headline"></a></h3>
<p>Now that we have added a dataset, we can configure visualizations. The <code class="docutils literal notranslate"><span class="pre">.add_view(dataset,</span> <span class="pre">component_type)</span></code> method adds a view (i.e. visualization or controller component) to the configuration.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">Component</span></code> enum class (which we have imported as <code class="docutils literal notranslate"><span class="pre">cm</span></code> here) can be used to fill in the <code class="docutils literal notranslate"><span class="pre">component_type</span></code> parameter.</p>
<p>For convenience, the <code class="docutils literal notranslate"><span class="pre">SCATTERPLOT</span></code> component type takes the extra <code class="docutils literal notranslate"><span class="pre">mapping</span></code> keyword argument, which specifies which embedding should be used for mapping cells to (x,y) points on the plot.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">scatterplot</span> <span class="o">=</span> <span class="n">vc</span><span class="o">.</span><span class="n">add_view</span><span class="p">(</span><span class="n">cm</span><span class="o">.</span><span class="n">SCATTERPLOT</span><span class="p">,</span> <span class="n">dataset</span><span class="o">=</span><span class="n">dataset</span><span class="p">,</span> <span class="n">mapping</span><span class="o">=</span><span class="s2">&quot;UMAP&quot;</span><span class="p">)</span>
<span class="n">cell_sets</span> <span class="o">=</span> <span class="n">vc</span><span class="o">.</span><span class="n">add_view</span><span class="p">(</span><span class="n">cm</span><span class="o">.</span><span class="n">CELL_SETS</span><span class="p">,</span> <span class="n">dataset</span><span class="o">=</span><span class="n">dataset</span><span class="p">)</span>
<span class="n">genes</span> <span class="o">=</span> <span class="n">vc</span><span class="o">.</span><span class="n">add_view</span><span class="p">(</span><span class="n">cm</span><span class="o">.</span><span class="n">GENES</span><span class="p">,</span> <span class="n">dataset</span><span class="o">=</span><span class="n">dataset</span><span class="p">)</span>
<span class="n">heatmap</span> <span class="o">=</span> <span class="n">vc</span><span class="o">.</span><span class="n">add_view</span><span class="p">(</span><span class="n">cm</span><span class="o">.</span><span class="n">HEATMAP</span><span class="p">,</span> <span class="n">dataset</span><span class="o">=</span><span class="n">dataset</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="4.4.-Define-the-visualization-layout">
<h3>4.4. Define the visualization layout<a class="headerlink" href="#4.4.-Define-the-visualization-layout" title="Permalink to this headline"></a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">vc.layout(view_concat)</span></code> method allows us to specify how our views will be arranged in the layout grid in the widget. The <code class="docutils literal notranslate"><span class="pre">|</span></code> and <code class="docutils literal notranslate"><span class="pre">/</span></code> characters are magic syntax for <code class="docutils literal notranslate"><span class="pre">hconcat(v1,</span> <span class="pre">v2)</span></code> and <code class="docutils literal notranslate"><span class="pre">vconcat(v1,</span> <span class="pre">v2)</span></code>, respectively.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">vc</span><span class="o">.</span><span class="n">layout</span><span class="p">((</span><span class="n">scatterplot</span> <span class="o">|</span> <span class="n">cell_sets</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">heatmap</span> <span class="o">|</span> <span class="n">genes</span><span class="p">));</span>
</pre></div>
</div>
</div>
</section>
</section>
<section id="5.-Launch-the-web-application">
<h2>5. Launch the web application<a class="headerlink" href="#5.-Launch-the-web-application" title="Permalink to this headline"></a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">vc.web_app()</span></code> method serves the processed data locally and opens a web browser to <code class="docutils literal notranslate"><span class="pre">http://vitessce.io/?url={config_as_json}</span></code></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">vc</span><span class="o">.</span><span class="n">web_app</span><span class="p">()</span>
</pre></div>
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020, Gehlenborg Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>